Add TLS support, gated by a feature flag

gferon opened the pull request:
Following up on my earlier attempt in MR #244.

TODO:
- [x] Implement TLS for synchronous client with [native-tls](https://crates.io/crates/native-tls)
- [x] Implement TLS with [tokio-tls](https://crates.io/crates/tokio-tls)
- [x] Implement TLS with [async-native-tls](https://crates.io/crates/async-native-tls)
- [x] Add testing facilities for `redis+tls`

Resolves issue #241

Marwes commented on src/connection.rs:
```diff
@@ -111,14 +132,39 @@ impl IntoConnectionInfo for String {
+    let addr = if url.scheme() == "redis+tls" {
+        #[cfg(feature = "tls")]
+        {
+            if url.fragment() == Some("insecure") {
```
Nit: Could do

```rust
let insecure = url.fragment() == Some("insecure");
ConnectionAddr::TcpTls {
                    host,
                    port,
                    insecure,
                }
```

Marwes replied:
Should any fragment other than `insecure` and `None` be an error?

gferon replied:
> Should any fragment other than `insecure` and `None` be an error?

I guess it makes sense, I just pushed a new commit to address both of your comments.

gferon replied:
@Marwes I guess I should also add something in the documentation about the new feature flag?

gferon commented:
@Marwes I guess I should also add something in the documentation about the new feature flag?

Marwes commented:
There doesn't seem to be a place where features are documented atm (though it would be nice, problably in the top level docs in src/lib.rs).

Adding some/a test for TLS, even without certificate validation would be useful.

gferon commented:
> There doesn't seem to be a place where features are documented atm (though it would be nice, problably in the top level docs in src/lib.rs).
> 
> Adding some/a test for TLS, even without certificate validation would be useful.

Alright, I'll figure something out later then.

gferon commented:
@Marwes could you try to help me figure out what's going on with the test suite? It's most probably linked to my change, but I can't figure it out right now.

badboy commented on src/aio.rs:
```diff
@@ -52,7 +59,14 @@ use crate::aio_async_std;
     /// Performs a TCP connection
     async fn connect_tcp(socket_addr: SocketAddr) -> RedisResult<ActualConnection>;
-    /// Performans an UNIX connection
+    // Performs a TCP TLS connection
```
```suggestion

    // Performs a TCP TLS connection
```

badboy replied:
Let's unclutter this a bit by adding newlines.

badboy commented on src/aio.rs:
```diff
@@ -52,7 +59,14 @@ use crate::aio_async_std;
+        socket_addr: SocketAddr,
+        insecure: bool,
+    ) -> RedisResult<ActualConnection>;
+    /// Performs a UNIX connection
```
```suggestion

    /// Performs a UNIX connection
```

badboy commented on src/aio.rs:
```diff
@@ -73,6 +90,27 @@ mod tokio_aio {
+        ) -> RedisResult<ActualConnection> {
+            let tls_connector: tokio_tls::TlsConnector = if insecure {
+                TlsConnector::builder()
+                    .danger_accept_invalid_certs(true)
```
That seems ... dangerous and I'd rather not set dangerous configuration by default.
Reducing the security of the TLS connection should be up to the user, not the library.

badboy replied:
Oh only now saw that this is behind the insecure flag. Ignore my comment then.

gferon replied:
Yeah, unfortunately it's a little hard to expose this `TlsConnector` to the user. We could figure it out, though.

badboy commented on src/aio.rs:
```diff
@@ -87,13 +125,21 @@ pub(crate) enum ActualConnection {
     #[cfg(feature = "tokio-comp")]
     TcpTokio(TcpStreamTokio),
+    /// Represents a Tokio TLS encrypted TCP connection
+    #[cfg(feature = "tokio-comp")]
+    #[cfg(feature = "tls")]
```
Can we combine these feature selects into a single line? I assume this would be:

```suggestion
    #[cfg(any(feature = "tokio-comp", feature = "tls"))]
```

badboy replied:
Wait, isn't this what we have the `tokio-tls-comp` feature for? Gosh, we're getting too many features :D

badboy replied:
Guess that applies to all usage of similar things below, also for `async-std-tls-comp`

badboy commented on src/aio.rs:
```diff
@@ -87,13 +125,21 @@ pub(crate) enum ActualConnection {
     #[cfg(feature = "async-std-comp")]
     TcpAsyncStd(aio_async_std::TcpStreamAsyncStdWrapped),
+    /// Represents an Async_std TLS encrypted TCP connection.
+    #[cfg(feature = "async-std-comp")]
+    #[cfg(feature = "tls")]
```
```suggestion
    #[cfg(any(feature = "async-std-comp", feature = "tls"))]
```

badboy commented on src/aio.rs:
```diff
@@ -390,10 +460,27 @@ async fn connect_simple<T: Connect>(
+        }
+
+        #[cfg(not(feature = "tls"))]
+        ConnectionAddr::TcpTls { .. } => {
```
We shouldn't even have this variant if it's not compiled with `tls`, right?

gferon replied:
well, it's like the `unix` one, it's mostly because we want to parse the address string and return an error if the user didn't configure the crate properly for their usage.

badboy commented on src/aio_async_std.rs:
```diff
@@ -1,5 +1,9 @@
 use crate::aio::{ActualConnection, Connect};
 use crate::types::RedisResult;
+#[cfg(feature = "tls")]
+use async_native_tls::TlsConnector;
+#[cfg(feature = "tls")]
+use async_native_tls::TlsStream;
```
```suggestion
use async_native_tls::{TlsConnector, TlsStream};
```

badboy commented on src/connection.rs:
```diff
@@ -22,7 +25,7 @@ static DEFAULT_PORT: u16 = 6379;
     match url::Url::parse(input) {
         Ok(result) => match result.scheme() {
-            "redis" | "redis+unix" | "unix" => Ok(result),
+            "redis" | "redis+tls" | "redis+unix" | "unix" => Ok(result),
```
I remember that some clients did use `rediss:` for secure connections (see also [the iana spec](https://www.iana.org/assignments/uri-schemes/prov/redis)).

Did you look what other clients are currently using?
(I'm ok with `redis+tls`, but if `rediss:` is used widely we maybe should have that as well)

gferon replied:
I merely copied `redis+unix` for consistency, but I guess the IANA spec is a much better source of truth.

badboy commented on src/connection.rs:
```diff
@@ -263,6 +318,67 @@ impl ActualConnection {
+            ConnectionAddr::TcpTls { .. } => {
+                fail!((
+                    ErrorKind::InvalidClientConfig,
+                    "Cannot connect to TCP with TLS without the tls feature"
```
Oh I see. We already do this for unix socket as well, so that explains why it is happening further above as well. I'm ok with that then.

Terkwood commented:
Curious about the build errors here.  The summary text says the were stopped, but the logs were a little bit harder to understand -- did someone intervene to stop the builds, or did they just idle for a long time and then fail?

Marwes commented:
travis stops the build automatically if no output has been received for 10 minutes. So there is a bug in here somewhere.

Terkwood commented on src/connection.rs:
```diff
@@ -38,6 +41,22 @@ pub fn parse_redis_url(input: &str) -> Result<url::Url, ()> {
+        /// You should think very carefully before you use this method. If hostname
+        /// verification is not used, any valid certificate for any site will be
+        /// trusted for use from any other. This introduces a significant
+        /// vulnerability to man-in-the-middle attacks.
```
:+1:

Terkwood commented:
Potentially related, in the travis log: https://travis-ci.org/github/mitsuhiko/redis-rs/jobs/693945209#L253

`Warning: The home dir /var/run/stunnel4 you specified can't be accessed: No such file or directory`

But probably benign.  I worry that something related to travis _might_ be the issue here, not the change set.  I'll pull down the repo and try running tests on my own machine.

Terkwood commented:
On my local machine (a linux laptop built in approximately 1433 AD), I was able to get the tests to pass:

```sh
REDISRS_SERVER_TYPE=tcp RUST_TEST_THREADS=1 cargo test --no-default-features --tests -- --nocapture
```

This is the same spot that failed in the travis build.  So I'm not sure there's anything wrong with the change set, and I'm a little bit more suspicious of the travis build's config.  Too -- and  I  haven't dug into this -- some of the other builds attached to the last commit, here, _passed_.  

Just out of sheer masochism, I wonder if re-triggering the builds causes different results than the last run.

Terkwood commented:
There have been changes since then, but previous iterations of this work got 100% green-lights on Travis:

https://travis-ci.org/github/mitsuhiko/redis-rs/builds/669753751   

https://github.com/mitsuhiko/redis-rs/pull/305/commits/789b3086e9492f9a6ae36f3118d63f61641e4e3c

Terkwood commented:
Someone has kindly kicked off the build for the latest commit:  https://travis-ci.org/github/mitsuhiko/redis-rs/builds/693945206

Thank you!

Terkwood commented:
Seems to be hanging at the same spot (test_getset). Run hasn't terminated yet...

Terkwood commented:
Failures on all of the non-lint builds, but not in the same places. 🔬

gferon commented:
I'm thinking it could be due to some stunnel processes that aren't killed properly, even though I don't see why the OS in the VM that Travis launches would behave differently than a container of the same distribution (which I tried locally)... Could we try to run the CI on master multiple times in a row to see if it somehow exhibits the same symptoms of flakiness?

Terkwood commented:
Based on work in #319, I'm relatively confident master doesn't have this issue. The only build failures I saw were routine things like lint errors.

Terkwood commented:
(I also don't have access to trigger the build on master 😁) If someone else is willing to do so it certainly won't hurt. Beyond that, I'm not sure what to suggest here. This is a tricky one to debug!

Terkwood commented:
Oh:You can  open a trivial fork with a one line PR, as a draft, and you'll trigger a build

Terkwood commented:
https://github.com/mitsuhiko/redis-rs/pull/337

testing with this PR

gferon commented:
Thanks for all the details, it 99.98% sounds like there's something up with my change (my gut feeling would be that it's linked with the test code). I'm back in front of a computer in 48h, I will try to reproduce the problem locally! I wouldn't want anybody spending too much time figuring out what I broke 😅

Terkwood commented:
No worries! This is an important feature. Initial checks in #337 show that master is clean

gferon commented:
I managed to reproduce the problem :tada: I'll post the result of the investigation soon.

gferon commented:
I was excited when I saw the update to redis 6.0, and I was too fast... the build from the `ppa:chris-lea/redis-server` isn't built with TLS support :crying_cat_face:

Terkwood commented:
😔 Thanks for looking into it... Worth a try

gferon commented:
I finally figured it out.

I introduced the issue in my PR when I changed the match arm in `tests/support/mod.rs` (see the [fix that restores this code to its original state](https://github.com/mitsuhiko/redis-rs/pull/305/commits/d331d20e5a6d5bdc2638245cf2d966e2844c5d5c)).

Before my change, the `net2::TcpListener2` was dropped _before_ the redis-server is spawned. After I changed it, well... it didn't work anymore :boom:

I thought I knew enough about sockets... It turns out that `SO_REUSEADDR` on Linux does not behave exactly as I thought it would. But hey, I [learned a lot in the process!](https://stackoverflow.com/a/14388707)

@Terkwood I think it's finally ready to be merged.

Terkwood commented on examples/basic.rs:
```diff
@@ -154,7 +155,12 @@ fn do_redis_code() -> redis::RedisResult<()> {
     // at this point the errors are fatal, let's just fail hard.
-    match do_redis_code() {
+    let url = if env::args().nth(1) == Some("--tls".into()) {
+        "rediss://127.0.0.1:6380/#insecure"
```
👍🏼

Terkwood commented:
Huge progress!! 🏁🏁

badboy commented:
Awesome!
I'm going to land this.

@gferon Can you add some documentation in lib.rs for this in a follow up PR?

tibbon commented:
How does one use the TLS feature?

tibbon commented:
I'm getting this error: 

```
david@MacBook-Pro-2 crane % cargo run -- -tls
   Compiling redis v0.17.0
error[E0432]: unresolved import `async_native_tls`
 --> /Users/david/.cargo/registry/src/github.com-1ecc6299db9ec823/redis-0.17.0/src/aio_async_std.rs:4:5
  |
4 | use async_native_tls::{TlsConnector, TlsStream};
  |     ^^^^^^^^^^^^^^^^ use of undeclared crate or module `async_native_tls`

error[E0433]: failed to resolve: use of undeclared crate or module `tokio_tls`
   --> /Users/david/.cargo/registry/src/github.com-1ecc6299db9ec823/redis-0.17.0/src/aio.rs:100:32
    |
100 |             let tls_connector: tokio_tls::TlsConnector = if insecure {
    |                                ^^^^^^^^^ use of undeclared crate or module `tokio_tls`

error[E0599]: no variant or associated item named `TcpTlsTokio` found for enum `aio::ActualConnection` in the current scope
   --> /Users/david/.cargo/registry/src/github.com-1ecc6299db9ec823/redis-0.17.0/src/aio.rs:113:40
    |
113 |                 .map(ActualConnection::TcpTlsTokio)?)
    |                                        ^^^^^^^^^^^
    |                                        |
    |                                        variant or associated item not found in `aio::ActualConnection`
    |                                        help: there is a variant with a similar name: `TcpTokio`
...
125 | pub(crate) enum ActualConnection {
    | -------------------------------- variant or associated item `TcpTlsTokio` not found here

error[E0599]: no variant or associated item named `TcpTlsAsyncStd` found for enum `aio::ActualConnection` in the current scope
   --> /Users/david/.cargo/registry/src/github.com-1ecc6299db9ec823/redis-0.17.0/src/aio_async_std.rs:161:42
    |
161 |             .map(|con| ActualConnection::TcpTlsAsyncStd(TlsStreamAsyncStdWrapped(con)))?)
    |                                          ^^^^^^^^^^^^^^
    |                                          |
    |                                          variant or associated item not found in `aio::ActualConnection`
    |                                          help: there is a variant with a similar name: `TcpAsyncStd`
    |
   ::: /Users/david/.cargo/registry/src/github.com-1ecc6299db9ec823/redis-0.17.0/src/aio.rs:125:1
    |
125 | pub(crate) enum ActualConnection {
    | -------------------------------- variant or associated item `TcpTlsAsyncStd` not found here

error: aborting due to 4 previous errors
```

I've included these in my Cargo.toml:

```
redis = { version = "0.17.0" , features = ["tls"] }
native-tls = { version = "0.2" }
tokio-tls = { version = "0.3" }
async-native-tls = { version = "0.3" }
```

tibbon commented:
Seems to be that this is the magic line: `redis = { version = "0.17.0" , features = ["tls", "async-std-tls-comp", "tokio-tls-comp"] }
`

gferon commented:
> Seems to be that this is the magic line: `redis = { version = "0.17.0" , features = ["tls", "async-std-tls-comp", "tokio-tls-comp"] } `

You're right, I just pushed #382 to make sure those features are more discoverable.
